### ISSUE TITLE
CC0119: replace if statement by null coalescing operator (C#)

### ISSUE THREAD

**Issue by giggio** [code-cracker/code-cracker#771]
An if statement that tests an expression for null and assigns the value to a variable if not null and assigns some other value if the the expression is null can be replaced by the null coalescing operator

Before:

````C#

      if (A != null)
        C = A;
      else
        C = B;

      if (A != null)
        return A;
      else
        return B;

      if (a == null)
        C = B;
      else
        C = A;

      if (a == null)
        return B;
      else
        return A;

      if (a == null)
        Z = A ? B : C;
      else
        Z = D ? E : F;
````

After:

````c#
      C = A ?? B;

      return A ?? B;

      C = A ?? B;

      return A ?? B;

      Z = (A ? B : C) ?? (D ? E : F);
````

The code fix should be careful with [operator precedence](https://msdn.microsoft.com/en-us/library/ms173145.aspx), as showed on the last example, where the parenthesis is needed.

Diagnostic Id: `CC0119`
Category: `Style` 
Severity: `Info`

Same as #765, for C#.

**Comment by JoelRamosM** [comment]
@giggio I want start to contribute, can i start by this one?

**Comment by vs-krasnov** [comment]
@giggio @JoelRamosM  Do I understand correctly that no one works on this issue - https://huboard.com/code-cracker/code-cracker/#/issues/147906180 ? If it is still open - I'd like to to work on it.

**Comment by JoelRamosM** [comment]
@vs-krasnov i'm ok,  I canot put this task in my agenda.

**Comment by vs-krasnov** [comment]
@JoelRamosM Okay, please let me know if you need any help with it.

**Comment by giggio** [comment]
Sorry everyone, I've been busy, and could not answer before.
@JoelRamosM asked first, he can go, if he wants it.
@vs-krasnov If Joel does not answer within a week or does not want it, you can take the issue. Just comment here so we know.

**Comment by WinCPP** [comment]
Hi guys, 

Is this still up for grabs...? May I work on this or collaborate? Please do let me know.

Thanks, 
Mandar

**Comment by carloscds** [comment]
@WinCPP Go ahead!

**Comment by giggio** [comment]
Yes, I am assuming nobody finished it before, go ahead @WinCPP .

**Comment by WinCPP** [comment]
Hi,

Using the following regular expression for "find in all files in the solution" (C# files), I didn't get any replaceable occurrences. Is this work still pending or am I missing something...?

if[ \t]*\(\b(_\w+|[\w-[0-9_]]\w*)\b[ \t]*(==|!=)[ \t]*null\)(?!.*\breturn\b)

Note: the negative lookahead for 'return' was to eliminate many statements which where of the grammar
"if (...) return...;"
Appreciate your inputs. Also, I'm new to contributing on github; so apologies in case I'm doing something wrong.

Thanks.

**Comment by carloscds** [comment]
@WinCPP I didn´t understand your question. Are you try to find code source for analyzer ?

**Comment by WinCPP** [comment]
@carloscds sorry for not being clear. I didn't get the actual ask in this issue. First I thought that the existing code needs to be scanned to use null coalescing operator (a trivial task), as per this thread, hence the above comment. But now I'm wondering if this issue is to argument the code cracker functionality to suggest places where null coalescing operator could be used in the target source code. Essentially I'm trying to figure out what is to be done...

**Comment by WinCPP** [comment]
Please ignore typo... Mobile autocorrect changed 'augment' to 'argument'...

**Comment by carloscds** [comment]
@WinCPP It's a new feature for CodeCracker.

**Comment by WinCPP** [comment]
@carloscds  Cool thanks for clarification. This is interesting... sort of grammar tree evaluation and folding, if I get it right. I am interested to work on this.

**Comment by carloscds** [comment]
@WinCPP Go ahead!
